lib/cancan/ability.rb

```diff
@@ -181,11 +181,20 @@ def authorize!(action, subject, *args)
     def unauthorized_message(action, subject)
       keys = unauthorized_message_keys(action, subject)
       variables = { action: action.to_s }
-      variables[:subject] = (subject.class == Class ? subject : subject.class).to_s.underscore.humanize.downcase
+      variables[:subject] = translate_subject(subject)
       message = I18n.translate(keys.shift, variables.merge(scope: :unauthorized, default: keys + ['']))
       message.blank? ? nil : message
     end
 
+    def translate_subject(subject)
+      klass = (subject.class == Class ? subject : subject.class)
+      if klass.respond_to?(:model_name)
```

Comment:
in which case the class does not respond to `model_name` ? I see two cases here but only one test.

Reply:
> in which case the class does not respond to model_name ?

I think there are cases which the class does not respond to `model_name` because in following spec use Array as ability subject. So I wrote this method considering cases which doesn't respond to `model_name`.

https://github.com/CanCanCommunity/cancancan/blob/develop/spec/cancan/ability_spec.rb#L498

> I see two cases here but only one test.

Test case which does not respond to `model_name` has wrote already here:

https://github.com/CanCanCommunity/cancancan/blob/develop/spec/cancan/ability_spec.rb#L496